### ISSUE TITLE
Singleton Rapporteur.checker is not threadsafe

### ISSUE THREAD

**Issue by nevinera** [envylabs/rapporteur#17]
This showed up in our (puma-served) app as this error: `RuntimeError: can't add a new key into hash during iteration`, coming out of the `Hash.new` block inside `MessageList#initialize`. It's very rare, because it only happens when you are iterating `Rapporteur.checker.messages` (to output them in `/status.json` for example) while another request on the same process attempts to write to it.

The simple fix would be to use `Thread#thread_variable_set` and `Thread#thread_variable_get` to make the `Rapporteur.checker` a per-thread singleton.. but that's a bit messy (I'd be happy to implement it if you want, but it would definitely spoil the current aesthetic of your code).

If the checker doesn't *need* to be a singleton, generating a new Checker for every request would be the simplest solution. But if a singleton is necessary for performance reasons, one traditional solution would be to create a CheckerPool that handles concurrent requests for 'the' checker by checking one out (inside a Mutex), potentially (but seldom actually) creating a new Checker if none are available.

I'd be happy to submit a PR for whichever approach you'd prefer, assuming you agree with my understanding of the issue :-)

**Comment by nbibler** [comment]
Interesting. Good catch. I'll have to think about how to approach this and follow up.

Thanks for the report!

**Comment by nbibler** [comment]
@nevinera: I added a PR with a fix (I think) for the issue. If you're able, please test it and let me know if it fixes your issue. You'll have to use a git ref in your Gemfile (something like):

```ruby
gem "rapporteur', github: 'envylabs/rapporteur', branch: 'thread-safety' # or
gem "rapporteur', github: 'envylabs/rapporteur', ref: '747512359e9432caa84e46f958ecc07ec1ab0dd7'
```

**Comment by nbibler** [comment]
Closed with 7475123.

**Comment by nbibler** [comment]
The fix was released in v3.6.1. Thanks for the report and let me know if this doesn't fix your issue!
